src/capture-metrics.js

```diff
@@ -1,20 +1,59 @@
+import { _ } from './utils'
+
 export class CaptureMetrics {
-    constructor(capture) {
+    constructor(enabled, capture, getTime = () => new Date().getTime()) {
+        this.enabled = enabled
         this.capture = capture
+        this.getTime = getTime
         this.metrics = {}
+        this.requests = {}
     }
 
     incr(key, by = 1) {
-        if (this.capture) {
+        if (this.enabled) {
             key = `phjs-${key}`
             this.metrics[key] = (this.metrics[key] || 0) + by
         }
     }
 
     decr(key) {
-        if (this.capture) {
+        if (this.enabled) {
             key = `phjs-${key}`
             this.metrics[key] = (this.metrics[key] || 0) - 1
         }
     }
+
+    startRequest(payload) {
+        if (this.enabled) {
+            const requestId = _.UUID()
+
+            this.requests[requestId] = [this.getTime(), payload]
+
+            return requestId
+        }
+    }
+
+    finishRequest(requestId) {
+        if (this.enabled) {
+            const [startTime, payload] = this.requests[requestId]
+            payload['duration'] = this.getTime() - startTime
+            delete this.requests[requestId]
+            return payload
+        }
+    }
+
+    markRequestFailed(payload) {
+        if (this.enabled) {
+            this.capture('$capture_failed_request', payload)
```

Comment:
If we're capturing this at unload wouldn't we have better results by batching all failed requests in a single API call?